File: value/src/main/java/com/bnorm/barkeep/model/UserValue.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model;

import javax.annotation.Nullable;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableSet;

@AutoValue
public abstract class UserValue implements User {

  public static Builder builder() {
    return new AutoValue_UserValue.Builder();
  }

  @Redacted
  @Nullable
  @Override
  public abstract String getPassword();

  @Override
  public abstract ImmutableSet<Bar> getBars();

  @Override
  public abstract ImmutableSet<Book> getBooks();

  @Override
  public abstract ImmutableSet<Recipe> getRecipes();

  @AutoValue.Builder
  public abstract static class Builder {

    public abstract UserValue build();

    public abstract Builder setId(Long id);

    public abstract Builder setUsername(String username);

    public abstract Builder setPassword(String password);

    public abstract Builder setDisplayName(String displayName);

    public abstract Builder setEmail(String email);

    abstract ImmutableSet.Builder<Bar> barsBuilder();

    public Builder addBar(Bar bar) {
      barsBuilder().add(bar);
      return this;
    }

    abstract ImmutableSet.Builder<Book> booksBuilder();

    public Builder addBook(Book book) {
      booksBuilder().add(book);
      return this;
    }

    abstract ImmutableSet.Builder<Recipe> recipesBuilder();

    public Builder addRecipe(Recipe recipe) {
      recipesBuilder().add(recipe);
      return this;
    }
  }
}

File: model-value/src/main/kotlin/com/bnorm/barkeep/model/value/UserValue.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model.value

import com.bnorm.barkeep.model.Bar
import com.bnorm.barkeep.model.Book
import com.bnorm.barkeep.model.Recipe
import com.bnorm.barkeep.model.User
import com.bnorm.barkeep.model.UserSpec
import com.squareup.moshi.FromJson
import com.squareup.moshi.ToJson

data class UserSpecValue(override val username: String?,
                         override val password: String? = null,
                         override val displayName: String? = null,
                         override val email: String? = null) : UserSpec

data class UserValue(override val id: Long = -1,
                     override val username: String?,
                     override val password: String? = null,
                     override val displayName: String? = null,
                     override val email: String? = null,
                     override val bars: Set<Bar> = emptySet(),
                     override val books: Set<Book> = emptySet(),
                     override val recipes: Set<Recipe> = emptySet()) : User

object UserValueAdapter {
  @ToJson fun toJson(user: UserSpec): UserSpecValue {
    return UserSpecValue(user.username,
                         user.password,
                         user.displayName,
                         user.email)
  }

  @FromJson fun fromJson(json: UserSpecValue): UserSpec {
    return json
  }

  @ToJson fun toJson(user: User): UserValue {
    return UserValue(user.id,
                     user.username,
                     user.password,
                     user.displayName,
                     user.email,
                     user.bars,
                     user.books,
                     user.recipes)
  }

  @FromJson fun fromJson(json: UserValue): User {
    return json
  }
}
